pkg/runtime/adoption_reconciler_test.go

```diff
@@ -0,0 +1,505 @@
+	k8sobj "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlrtzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	ackv1alpha1 "github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1"
+	k8srtmocks "github.com/aws-controllers-k8s/runtime/mocks/apimachinery/pkg/runtime"
+	ctrlrtclientmock "github.com/aws-controllers-k8s/runtime/mocks/controller-runtime/pkg/client"
+	ackmocks "github.com/aws-controllers-k8s/runtime/mocks/pkg/types"
+	ackcfg "github.com/aws-controllers-k8s/runtime/pkg/config"
+	ackmetrics "github.com/aws-controllers-k8s/runtime/pkg/metrics"
+	ackrt "github.com/aws-controllers-k8s/runtime/pkg/runtime"
+	ackrtcache "github.com/aws-controllers-k8s/runtime/pkg/runtime/cache"
+	acktypes "github.com/aws-controllers-k8s/runtime/pkg/types"
+)
+
+const(
+	Namespace = "default"
+	Name = "adoptedRes"
+)
+
+// Helper functions for tests
+
+func mockReconciler() (acktypes.AdoptedResourceReconciler, *ctrlrtclientmock.Client ,*ctrlrtclientmock.Reader) {
+	zapOptions := ctrlrtzap.Options{
+		Development: true,
+		Level:       zapcore.InfoLevel,
+	}
+	fakeLogger := ctrlrtzap.New(ctrlrtzap.UseFlagOptions(&zapOptions))
+	cfg := ackcfg.Config{}
+	metrics := ackmetrics.NewMetrics("bookstore")
+
+	sc := &ackmocks.ServiceController{}
+	rmfactory := ackmocks.AWSResourceManagerFactory{}
+	rmFactoryMap := make(map[string]acktypes.AWSResourceManagerFactory)
+	rmFactoryMap["services.k8s.aws"] = &rmfactory
+	sc.On("GetResourceManagerFactories").Return(rmFactoryMap)
+	kc := &ctrlrtclientmock.Client{}
+	apiReader := &ctrlrtclientmock.Reader{}
+	return ackrt.NewAdoptionReconcilerWithClient(
+		sc,
+		fakeLogger,
+		cfg,
+		metrics,
+		ackrtcache.Caches{},
+		kc,
+		apiReader,
+		), kc, apiReader
+}
+
+func mockDescriptorAndAWSResource() (*ackmocks.AWSResourceDescriptor, *ackmocks.AWSResource) {
+	des := &ackmocks.AWSResourceDescriptor{}
+	emptyRuntimeObject := &k8srtmocks.Object{}
+	res := &ackmocks.AWSResource{}
+	des.On("EmptyRuntimeObject").Return(emptyRuntimeObject)
+	des.On("ResourceFromRuntimeObject", emptyRuntimeObject).Return(res)
+	return des, res
+}
+
+func mockManager() *ackmocks.AWSResourceManager {
+	return &ackmocks.AWSResourceManager{}
+}
+
+func setupMockClient(kc *ctrlrtclientmock.Client, statusWriter *ctrlrtclientmock.StatusWriter, ctx context.Context, adoptedRes *ackv1alpha1.AdoptedResource) {
+	kc.On("Status").Return(statusWriter)
+	statusWriter.On("Patch", ctx, adoptedRes, mock.AnythingOfType("*client.mergeFromPatch")).Return(nil)
+	kc.On("Patch", ctx, adoptedRes, mock.AnythingOfType("*client.mergeFromPatch")).Return(nil)
+}
+
+func setupMockAwsResource(res *ackmocks.AWSResource, adoptedRes *ackv1alpha1.AdoptedResource) {
+	res.On("SetIdentifiers", adoptedRes.Spec.AWS).Return(nil)
+	res.On("RuntimeObject").Return(&k8srtmocks.Object{})
+	res.On("SetObjectMeta", mock.AnythingOfType("ObjectMeta")).Run(func(args mock.Arguments) {})
```

Comment:
Empty mock method? Should this be actually setting the meta for it to mock correctly?

Reply:
No, res is mock object. We cannot save anything there.
This empty method is needed to set expected behavior for this call otherwise an ugly warning is printed that mock does not know what to do on this method call.
There is also no return type on `SetObjectMeta`, so cannot simply return `nil`